test/e2e/etcd_multi_node_test.go

```diff
@@ -0,0 +1,441 @@
+
+		ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+		if *etcd.Status.ObservedGeneration <= oldEtcdObservedGeneration {
+			return fmt.Errorf("waiting for etcd %q rolling update to complete", etcd.Name)
+		}
+		return nil
+	}, timeout*3, pollingInterval).Should(BeNil())
+}
+
+// hibernateAndCheckEtcd scales down etcd replicas to zero and ensures
+func hibernateAndCheckEtcd(ctx context.Context, cl client.Client, logger logr.Logger, etcd *v1alpha1.Etcd) {
+	ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+	etcd.Spec.Replicas = 0
+	etcd.SetAnnotations(
+		map[string]string{
+			v1beta1constants.GardenerOperation: v1beta1constants.GardenerOperationReconcile,
+		})
+	ExpectWithOffset(1, cl.Update(ctx, etcd)).ShouldNot(HaveOccurred())
+	logger.Info("Waiting to hibernate")
+
+	logger.Info("Checking statefulset")
+	Eventually(func() error {
+		sts := &appsv1.StatefulSet{}
+		ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), sts)).To(Succeed())
+
+		if sts.Status.ReadyReplicas != 0 {
+			return fmt.Errorf("sts %s not ready", etcd.Name)
+		}
+		return nil
+	}, timeout*3, pollingInterval).Should(BeNil())
+
+	logger.Info("Checking etcd")
+	Eventually(func() error {
+		etcd := getEmptyEtcd(etcd.Name, namespace)
+		err := cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)
+		if err != nil {
+			return err
+		}
+
+		if etcd != nil && etcd.Status.Ready != nil {
+			if *etcd.Status.Ready != true {
+				return fmt.Errorf("etcd %s is not ready", etcd.Name)
+			}
+		}
+
+		if etcd.Status.ClusterSize == nil {
+			return fmt.Errorf("etcd %s cluster size is empty", etcd.Name)
+		}
+		// TODO: uncomment me once scale down is supported,
+		// currently ClusterSize is not updated while scaling down.
+		// if *etcd.Status.ClusterSize != 0 {
+		// 	return fmt.Errorf("etcd %q cluster size is %d, but expected to be 0",
+		// 		etcdName, *etcd.Status.ClusterSize)
+		// }
+
+		for _, c := range etcd.Status.Conditions {
+			if c.Status != v1alpha1.ConditionUnknown {
+				return fmt.Errorf("etcd %s status condition is %q, but expected to be %s ",
+					etcd.Name, c.Status, v1alpha1.ConditionUnknown)
+			}
+		}
+
+		return nil
+	}, timeout*3, pollingInterval).Should(BeNil())
+	logger.Info("etcd is hibernated")
+}
+
+// wakeupEtcd scales up etcd replicas to 3 and ensures etcd cluster with 3 replicas is ready.
+func wakeupEtcd(ctx context.Context, cl client.Client, logger logr.Logger, etcd *v1alpha1.Etcd) {
+	ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+	etcd.Spec.Replicas = multiNodeEtcdReplicas
+	updateAndCheckEtcd(ctx, cl, logger, etcd)
+}
+
+// updateAndCheckEtcd updates the given etcd obj in the Kubernetes cluster.
+func updateAndCheckEtcd(ctx context.Context, cl client.Client, logger logr.Logger, etcd *v1alpha1.Etcd) {
+	sts := &appsv1.StatefulSet{}
+	ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), sts)).To(Succeed())
```

Comment:
This helper func is called in tests directly as well as by other helper funcs. Thus, there isn't always an offset of `1`.

Reply:
Fixed, `updateAndCheckEtcd` func is called in tests only.